### ISSUE TITLE
Feature: Configurable Button/Key Mapping

### ISSUE THREAD

**Issue by jahnf** [jahnf/Projecteur#69]
It must be possible to map input events (and some combinations like double tap) 
from the device to either
* certain application actions (e.g. switch zoom on/off, change spot shape ...)
* or to specified other keys/events (e.g. some device button maps to the Enter-key)

This should be configurable by the user (via config file and/or UI).

**Comment by maehne** [comment]
In issue #76, I [posted about the following idea](https://github.com/jahnf/Projecteur/issues/76#issuecomment-633223842), which is related to this feature request:

> If you want to use the spotlight for explanations over a longer time, it can be tiresome to have to leave the thumb on the button. How about an option where with one click on the device button the spotlight is turned on and with another click the spotlight is turned off? A kind of spotlight locking function might be even more intuitive: I the spotlight button got pressed for a configurable amount of time (e.g. 3 or 5 sec), it remains on -- even if the button is released. Only a second press on the button will turn it off again.

@jahnf [replied](https://github.com/jahnf/Projecteur/issues/76#issuecomment-633225805):

> Things like that will be possible with the button mapping functionality I am working on for the next release - maybe not all kind of combinations right from the start but Projecteur will get there - at least that is my plan. So if the user wants to, all buttons (or button sequences) they can be mapped to either a defined key-press or some action (e.g. toggle the spot??)
> Can't promise a date on this yet though.

**Comment by jahnf** [comment]
**Button mapping is here!**  :tada: 
So for everybody that is interested in it, please give it a try :wrench: :hammer: and test it.
Feedback and bug reports are welcome as always. Suggestions for generic actions (still to be implemented) for mapping are also highly appreciated.
 
A lot of custom widgets were implemented for this and there is still work to do but:

_**The following is already implemented**_
- [x]  Map any button (sequence) to (almost) any Key-Combination-Sequence
- [x]  Edit the device and key combinations comfortably within the GUI

_**The following is still to be implemented**_
- [ ]  Mapping of device buttons to a generic action (e.g. toggle zoom..., ???)
- [ ]  Highlight duplicate device key combinations (currently one of the duplicates is silently ignored)
- [ ]  Better visualization of device input sequence (e.g. combine button press/release)
- [ ]  Fix visualization of Key combination involving `AltGr`-key (although QKeySequenceEdit suffers from the same problem)
- [ ]  ... all the things that I could not think of right now.

**Comment by mayanksuman** [comment]
The key combination like Alt+Tab can not be added to mapper currently.

**Comment by jahnf** [comment]
> The key combination like Alt+Tab can not be added to mapper currently.

True, shortcuts like that are caught by GNOME or KDE (or any other Desktop environment)- so a keypress like that will not even arrive in the application itself. I don't have any idea if it is possible.

**Comment by mayanksuman** [comment]
To capture those kind of key combinations, can we grab the keyboard input when `Mapped Key(s)` is under edit?

**Comment by jahnf** [comment]
Yes, but I would **not** do that (grabbing and hijacking the users keyboard). Also: first of all you would need to know which input device the user is using to input the shortcut (could be any of multiple keyboards connected). Also you would need to have read access to the keyboard input device or be running as root user.

**Comment by jahnf** [comment]
The implementation allows now for more generic actions mapped to button input,
(this also means that if you had already button mappings set up they are lost - since the format for storing and loading internally changed) 

One other action is already implemented: 
* "Cycle Presets" : You can cycle through all your defined presets with a button (combination). Basically it's the "change presenter mode" functionality from the Logitech software on steroids.
